만국박람회 [STEP3] Cheshire, Claire

hano-ios opened the pull request:
@wonhee009 안녕하세요!
STEP3 PR입니다!

STEP 3을 하며 변경점은
- Label, Button 다이나믹 구현
- MainView AttributedString 구현
- 첫화면(MainView) 세로 고정

입니다.

# Label, Button 다이나믹 타입 구현
Label의 font를 모두 preferredFont(.title1, .body 등)으로 맞춰주고
adjustsFontForContentSizeCategory 속성을 true로 해주어 다이나믹 타입을 구현했습니다.
또한 Button의 titleLabel에도 동일하게 구현하였고, 
label이 버튼 영역을 벗어나는 경우를 위해 button과 titleLabel에 마진을 두어 해결했습니다.

# AttributedString
방문객: 00명과 같은 MainView 요소에서 특정 문자열만 다른 폰트로 하고 싶을 때
원래는 Label 두개로 구현했지만, AttributedString을 이용해서 스택뷰 없이 Label 하나로 구현할 수 있었습니다!

# MainView 세로고정
저희는 첫화면을 세로 고정, 나머지 화면은 자유 회전으로 구현했습니다.
그래서 각 화면의 동작은 구현했고, 기기를 가로로 두고 앱을 시작하면 MainView는 세로로 고정된 채로 기기만 회전되어 있는 것을 확인했습니다.
그리고 다음 화면으로 넘어갔을 때, 자동으로 화면이 회전되었으면 좋겠다고 생각하여 여러 시도를 해봤지만 해결하지는 못했습니다.

시도했던 것은 다음과 같습니다
- 두번째 화면에서 직접 회전
- 첫번째 화면 viewWillDisappear에서 기기 회전 넘겨서 회전
- UIDeviceOrientation을 이용한 회전 판별
- view.bounds의 가로/세로 길이를 통한 화면 회전 판별

여기서 의문이 든 점은
- 두번째 화면에서 회전 방향에 대한 출력을 해보면 정확한 값이 나오는데, 회전 함수를 호출하면 왜 회전하지 않는지..

코드는 다음과 같습니다.
```Swift
// EntryListViewController
AppDelegate.AppUtility.lockOrientation(UIInterfaceOrientationMask.all, andRotateTo: UIInterfaceOrientation(rawValue: UIDevice.current.orientation.rawValue)!)
```
```Swift
// AppDelegate
    struct AppUtility {
        static func lockOrientation(_ orientation: UIInterfaceOrientationMask) {
            if let delegate = UIApplication.shared.delegate as? AppDelegate {
                delegate.orientationLock = orientation
            }
        }

        static func lockOrientation(_ orientation: UIInterfaceOrientationMask, andRotateTo rotateOrientation:UIInterfaceOrientation) {
            self.lockOrientation(orientation)
            UIDevice.current.setValue(rotateOrientation.rawValue, forKey: "orientation")
        }
    }
```
이부분에 가장 많은 고민을 했었는데.. 혹시 해결방법을 아신다면 힌트 부탁드립니다..!

기타 피드백이나 고칠 사항 있으면 말씀해주시면 감사하겠습니다!!
항상 꼼꼼하고 친절한 피드백 감사드립니다 !

Commit:
feat: 폰트 다이나믹, 메인뷰 AttributedString 추가

폰트가 다이나믹 타입으로 변하게 구현
AttributedString을 이용해서 폰트 스택뷰 제거

Commit:
feat: 텍스트 다이나믹 타입 설정, 화면 고정

첫화면 세로 고정
모든 텍스트 다이나믹 설정

Commit:
feat: 메인화면 버튼 수정
메인화면 버튼 크기 자동조절되도록 수정
셀 이미지 높이 수정

Commit:
fix: MainView 스크롤뷰 컨테이너뷰 추가

warning이 뜨는 문제 해결 위해 스크롤 뷰 내부에 컨테이너뷰를 만들어
스크롤뷰 컨텐츠에 대한 사이즈를 잡아줌

Commit:
feat: 보너스 스테이지 구현(UIDevice)

entryDetailView에서 axis, image를 회전에 따라 다르게 함
EntryDetailViewController에서 스택뷰와 이미지 뷰 추가,
회전에 따라 다르게 구성함

Commit:
chore: isHidden 이용

imageView size를 0으로 두는 방식에서
isHidden을 이용하도록 수정

wonhee009 commented:
타겟 브랜치 지정이 잘못된거 같습니다! 확인 부탁드려요🙏

hano-ios commented:
> 타겟 브랜치 지정이 잘못된거 같습니다! 확인 부탁드려요🙏

수정했습니다 감사합니다...ㅎㅎㅎ

Commit:
fix: 보너스 스테이지 회전 수정

회전해도 해당 화면에서 바로 바뀌지 않는 문제 수정
viewWillTransition 함수에서 화면 전환

Commit:
chore: changeViewLayout 함수 추가

중복되는 코드 제거 -> changeViewLayout 함수로 통합

hano-ios commented:
보너스 스테이지도 구현해서 push했습니다!!
@wonhee009
@yeokyeong46

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryDetailViewController.swift:
```diff
@@ -1,12 +1,35 @@
 import UIKit
 
 class EntryDetailViewController: UIViewController {
+    private let entryImage: UIImageView = {
+        let imageView = UIImageView()
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        imageView.contentMode = .scaleAspectFit
+        return imageView
+    }()
+    
+    private lazy var totalView: UIStackView = {
+        let stackView = UIStackView(arrangedSubviews: [entryImage, scrollView])
+        stackView.alignment = .center
+        stackView.distribution = .fill
+        return stackView
+    }()
+    
```
해당 UI Property들은 가로일때 보여주기 위해서 추가된거로 보여요.
(또한, 프로퍼티 추가로 landscapeConstraints, portraitConstraints 등도 추가된거 같구요.)
이렇게 추가 프로퍼티들을 사용하는게 아니라 기존의 detailView의 axis를 방향에 맞춰서 변경해주는 방향은 어떨까요?
(이때에는 이미지의 width, height를 모두 정의하는게 좋을거 같습니다)

hano-ios replied:
이미지를 방향에 따라 superview 바깥으로 빼는 방식으로 다시 구현해보았습니다!! 확인 부탁드립니당

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryDetailViewController.swift:
```diff
@@ -21,14 +44,42 @@ class EntryDetailViewController: UIViewController {
+        changeViewLayout(size: size)
+    }
+    
+    override func viewWillLayoutSubviews() {
+        super.viewWillLayoutSubviews()
+    }
```
따로 오버라이드 해 준 부분이 없는거 같은데 추가된 이유가 있을까요?

hano-ios replied:
아! 다른 부분 시도하다가 정의한 메서드였습니다. 제거하겠습니다!

wonhee009 commented on Expo1900/Expo1900/ViewController/MainViewController.swift:
```diff
@@ -29,12 +35,14 @@ class MainViewController: UIViewController {
+        AppDelegate.AppUtility.lockOrientation(UIInterfaceOrientationMask.portrait, andRotateTo: UIInterfaceOrientation.portrait)
+    }
+    
+    override func viewWillDisappear(_ animated: Bool) {
+        navigationController?.setNavigationBarHidden(false, animated: false)
+        AppDelegate.AppUtility.lockOrientation(UIInterfaceOrientationMask.all, andRotateTo: UIInterfaceOrientation.unknown)
     }
```
main -> list로 화면이 넘어갈때 가로로 회전이 되었으면 좋겠다로 이해했습니다.
```swift
AppDelegate.AppUtility.lockOrientation(UIInterfaceOrientationMask.all, andRotateTo: UIInterfaceOrientation.landscapeLeft)
```
이렇게 설정하면 list화면으로 이동했을때 가로로 화면이 회전되는데요.
이 부분이 의도하신 부분과 일치할까요?

hano-ios replied:
https://user-images.githubusercontent.com/24950485/149916425-7211e64d-7777-4156-b5e8-599aa523f0db.mov

저희가 코드를 이전 버전으로 잘못 커밋해서 설명이 조금 부족했던 것 같습니다...!
원래 코드는 다음과 같습니당
```Swift
AppDelegate.AppUtility.lockOrientation(UIInterfaceOrientationMask.all, andRotateTo: UIInterfaceOrientation(rawValue: UIDevice.current.orientation.rawValue)!)
```

ListView의 화면이 기기가 세로일 때에는 세로 화면으로, 기기가 가로 화면일 때에는 가로 화면으로 나왔으면 좋겠다고 생각해서 UIDevice.current.orientation.rawValue의 값으로 UIInterfaceOrientation을 바꿔주려 했는데, 막상 결과는 위의 영상처럼 나와서 질문 드렸습니다!

조금 헷갈리게 말씀드린 것 같아 다시 정리하자면,
1) 기기가 세로인 경우
MainView(세로 interface) -> ListView (세로 interface)
2) 기기가 가로인 경우
MainView(세로 interface, 기기는 가로모드) -> ListView(가로 interface)

으로 동작했으면 좋겠다고 생각했습니다.
라자냐가 말씀해주신 코드는 2번의 경우만 커버하는 것 같아서 조금 더 설명을 추가했는데 의미가 전달이 되었으면 좋겠습니당!

한번 다시 확인해주신다면 감사하겠습니다 ㅠㅠ

wonhee009 replied:
<img width="808" alt="image" src="https://user-images.githubusercontent.com/30456206/149922783-ef43cf1d-5c0e-4052-8e3e-17c8e4e8aee2.png">
이렇게 프린트문을 찍어보면
<img width="56" alt="image" src="https://user-images.githubusercontent.com/30456206/149923189-fa4cbc56-4cb7-4db5-872e-8ff2bd11a66a.png">
appear 시점과 disappear 시점의 rotationOrientation이 다르게 찍히고 있네요.
[공식 문서](https://developer.apple.com/documentation/uikit/uiinterfaceorientation/unknown)에 따르면 0은 unknown의 경우인데요.
orientation이 제대로 들어가지 않고 있네요ㅠ_ㅠ

hano-ios replied:
넵 다시 확인해보니 landscape에서 UIDevice.current.setValue를 하면서 문제가 생겼던 것 같습니다..
다른 코드로 수정해서 다시 commit했습니다!!

wonhee009 reviewed (changes requested):
스텝3 수고하셨습니다!
몇몇 코멘트와 달아두었으니 확인 부탁드려요!🙏

Commit:
fix: orientation 수정, DetailView 삭제

orientation 화면 전환시 회전하도록 수정
detailView 삭제 후 detailViewController에서 구현
imageView 재사용함

hano-ios commented:
수정해서 다시 커밋했습니다 !!!
@wonhee009  @yeokyeong46

wonhee009 commented on Expo1900/Expo1900/ViewController/EntryDetailViewController.swift:
```diff
@@ -66,18 +78,24 @@ class EntryDetailViewController: UIViewController {
-            detailView.hideImage(true)
+            entryImage.removeFromSuperview()
+            totalView.insertArrangedSubview(entryImage, at: 0)
             NSLayoutConstraint.activate(landscapeConstraints)
             NSLayoutConstraint.deactivate(portraitConstraints)
         }
         else {
             totalView.axis = .vertical
-            entryImage.isHidden = true
-            detailView.hideImage(false)
+            entryImage.removeFromSuperview()
+            detailView.insertArrangedSubview(entryImage, at: 0)
```
UI요소를 superView에서 remove, insert하는 방식을 사용하셨네요.
UI요소를 remove하는 작업은 비용이 높은 작업여서 이런 작업보다는 constraint를 활용하도록 권장하는데요.
entryImage를 방향에 따라 삭제/삽입하지 않고 `axis`만 변경해도 해결 가능할거 같습니다:)
